File: leetcode/78/78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> oneAns;
        ans.push_back(oneAns);
        for(int i = 1; i <= nums.size(); i++) {
            findOneAns(oneAns, 0, nums, i);
        }
        return ans;
    }

    void findOneAns(vector<int>& oneAns, int start, vector<int>& nums, int k) {
		for (; start <= nums.size() - k + oneAns.size(); start++) {
			oneAns.push_back(nums[start]);
			if (oneAns.size() == k) {
				ans.push_back(oneAns);
				oneAns.pop_back();
			}
			else {
				findOneAns(oneAns, start + 1, nums, k);
				oneAns.pop_back();
			}
		}
	}
private:
    vector<vector<int>> ans;
};

File: leetcode/78/78.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        int len = nums.length;
        for(int i = 0; i <= (int)Math.pow(2, len) - 1; i++) {
            addOneAns(nums, i, ans);
        }
        return ans;
    }

    private void addOneAns(int[] nums, int status, List<List<Integer>> ans) {
        List<Integer> list = new ArrayList<>();
        int len = nums.length;
        for(int i = 0; i < len; i++) {
            if(((status >>> i) & 1) == 1) {
                list.add(nums[len - i - 1]);
            }
        }
        ans.add(list);
    }
}